### PR TITLE
Non-native field multiplication optimization 2

### LINKED ISSUE #930: Foreign field multiplication optimization 2 - OPT2
To reduce the number of rows, the foreign field multiplication gate constrains the [Bound Addition](https://github.com/o1-labs/proof-systems/blob/master/book/src/rfcs/foreign_field_mul.md#bound-checks) for the quotient.  Normally constraining the quotient bound addition requires constraining two carry bits (lo and hi); however, we have [proven formally that the highest carry bit is always zero](https://github.com/o1-labs/proof-systems/blob/c76995aa9e4a708986211edacd0b6e75344ce2a2/kimchi/src/tests/foreign_field_mul.rs#L244).

This enables us to perform the following optimization
* Reduce the number of constraints by 1 (eliminate the boolean constraint for the high carry bit)
* Remove the high carry bit from the witness

### PR DESCRIPTION
**Pull request by jspada**
This needs to be merged after #924 

Closes #930

### CODE REVIEW

File: kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs [hunk 1]
```diff
@@ -135,28 +136,23 @@ fn compute_witness_variables<F: PrimeField>(
         carry0,
         carry1_lo,
         carry1_hi,
-        scaled_carry1_hi,
     ]
     .to_fields()
 }
 
 fn compute_bound_witness_variables<F: PrimeField>(
```
Comment (querolita) [inline kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs:143]: perhaps a better name could be compute_bound_witness_carry now

**Review by querolita — APPROVED** [review]
Perfect, this ended up so succinct 😃

File: book/src/rfcs/foreign_field_mul.md [hunk 2]
```diff
@@ -1115,16 +1164,14 @@ NB: the $f$ and $f'$ values are publicly visible in the gate coefficients.
...
-| 13         | $q'_{carry01}$                           |                  |
-| 14         | $q'_{carry2}$                            |                  |
-
-where $\mathsf{scaled}_{v_{11}} = 2^9 \cdot v_{11}$.
+| 7          | $v_{11}$ (scaled plookup)                |                  |
+| 8          | $v_0$                                    |                  |
+| 9          | $q_0$                                    |                  |
+| 10         | $q_1$                                    |                  |
+| 11         | $q_2$                                    |                  |
+| 12         | $q'_{carry01}$                           |                  |
```
[the remark this answers is not in the document]
Comment (jspada) [reply book/src/rfcs/foreign_field_mul.md:1172]: True.. but also I like to know that it corresponds to compact limb with subscript `01`.

File: book/src/rfcs/foreign_field_mul.md [hunk 3]
```diff
@@ -1174,20 +1221,17 @@ In total we require the following checks
...
+10. $v_1 = 2^{\ell} \cdot v_{11} + v_{10}$
+11. $2^{\ell} \cdot v_1 = v_0 + p_{11} + p_2 - r_2$
+12. $a_n \cdot b_n - q_n \cdot f_n = r_n$
+13. $q'_0 \in [0, 2^{\ell})$ `multi-range-check`
+14. $q'_1 \in [0, 2^{\ell})$ `multi-range-check`
+15. $q'_2 \in [0, 2^{\ell})$ `multi-range-check`
+16. $q'_{01} = q'_0 + 2^{\ell} \cdot q'_1$ `multi-range-check`
+17. $q'_{carry01} \in [0, 2)$
+18. $2^{2\ell} \cdot q'_{carry01} = s_{01} - q'_{01}$
+19. $q'_2 = s_2 + q'_{carry01}$
```
[the remark this answers is not in the document]
Comment (jspada) [reply book/src/rfcs/foreign_field_mul.md:1234]: Ty!

File: kimchi/src/circuits/polynomials/foreign_field_mul/circuitgates.rs [hunk 4]
```diff
@@ -334,45 +326,34 @@ where
...
-        // C14: Constrain q'_carry2 is boolean
-        constraints.push(quotient_bound_carry2.boolean());
-
-        // C15: Constrain that 2^L * q'_carry2 = s2 + q'_carry01 - q'2
-        constraints.push(
-            T::two_to_limb() * quotient_bound_carry2 - sum2 - quotient_bound_carry01
-                + quotient_bound2,
-        );
+        // C12: Constrain that q'_2 = s2 + q'_carry01
+        constraints.push(quotient_bound2 - sum2 - quotient_bound_carry01);
```
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/circuits/polynomials/foreign_field_mul/circuitgates.rs:356]: You are starting to convince me.

File: kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs (hunk 1 above)
Comment (jspada) [reply kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs:143]: Also true.

File: kimchi/src/circuits/polynomials/foreign_field_mul/circuitgates.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/circuits/polynomials/foreign_field_mul/circuitgates.rs:356]: ok, I'm convinced.

**Pull request merged**